backend/src/Database.hs

```diff
@@ -1,108 +1,111 @@
 module Database where
 
-import Control.Monad.Logger (MonadLogger, logErrorN, logInfoN)
-import Data.Time (Year)
+import AppConfig (AppM, Env (..), runAppLogger)
+import Control.Monad.Logger (LoggingT, MonadLogger, logInfoN)
 import Database.Esqueleto.Experimental
   ( Entity (..),
     Key,
     SqlPersistT,
     desc,
     from,
-    get,
     getBy,
     innerJoin,
     insert,
     insert_,
     on,
     orderBy,
     replace,
+    runSqlPool,
     select,
     selectOne,
     table,
     val,
     where_,
+    (&&.),
     (==.),
     (^.),
     type (:&) (..),
   )
-import Logging ((<>:))
+import Servant (ServerError, throwError)
 import Types.DataField (PlayerName)
 import Types.Database
   ( EntityField (..),
     GameReport (..),
     Key (..),
     Player (..),
     PlayerId,
-    PlayerStats (..),
+    PlayerStatsTotal (..),
+    PlayerStatsYear (..),
     RatingDiff,
     Unique (..),
     currentYear,
     defaultPlayerStats,
-    rolloverPlayerStats,
   )
 import Prelude hiding (get, on)
 
-getPlayerByName :: (MonadIO m, MonadLogger m) => PlayerName -> SqlPersistT m (Maybe (Entity Player))
-getPlayerByName name = getBy $ UniquePlayerName name
+type DBAction m = ExceptT ServerError (SqlPersistT m)
 
-insertPlayerIfNotExists :: (MonadIO m, MonadLogger m) => PlayerName -> SqlPersistT m (Key Player)
-insertPlayerIfNotExists name =
-  getPlayerByName name >>= \case
+runDb :: DBAction (LoggingT IO) a -> AppM a
+runDb dbAction = do
+  env <- ask
+  result <- liftIO . runAppLogger env.logger . runSqlPool (runExceptT dbAction) $ env.dbPool
+  case result of
+    Left err -> throwError err
+    Right a -> pure a
+
+getPlayerByName :: (MonadIO m, MonadLogger m) => PlayerName -> DBAction m (Maybe (Entity Player))
+getPlayerByName = lift . getBy . UniquePlayerName
```

Comment:
One little annoyance is that our functions now run in `ExceptT ServerError (SqlPersistT m)`, but `esqueleto`'s library functions still run in `SqlPersistT m`. To get around this, we need to use `lift` to lift the default monadic actions from `esqueleto`'s library, into our "exception handling" version of the monad.